rlberry/agents/dqn/pytorch.py

```diff
@@ -0,0 +1,104 @@
+import logging
+import torch
+from gym import spaces
+
+from rlberry.agents.common.models import size_model_config, model_factory, trainable_parameters
+from rlberry.agents.common.training import loss_function_factory, optimizer_factory
+from rlberry.agents.dqn.abstract import AbstractDQNAgent
+from rlberry.agents.dqn.memory import Transition
+from rlberry.utils.torch import choose_device
+
+logger = logging.getLogger(__name__)
+
+
+class DQNAgent(AbstractDQNAgent):
+    def __init__(self, env, config=None):
```

Comment:
For compatibility with the `rlberry.agents.Agent` class, I think this line should be:

```
def __init__(self, env, **config):
```

and I believe the rest of the code can remain unchanged, right?

Reply:
Yes, that is inherited from my personal way of handling configurations: a Configurable object should have all its config in a `config` dict, so that it can be saved to and loaded from a text file, in a generic way. See an example configuration here: https://github.com/eleurent/rl-agents/blob/master/scripts/configs/HighwayEnv/agents/DQNAgent/dqn.json

I'll see if I can adapt it to the current convention. But I don't really see how we can distinguish configurations (which are relevant to save/load and edit in a text format) from other run-specific variables (current episode, rewards, etc.). Using pickle does not seem very advisable here.

Reply:
What about doing the following:

```python
class SomeAgent(Agent, Configurable):
     def __init__(self, env, **config):
            Agent.__init__(self, env, **config)
            Configurable.__init__(self, config)
```

Reply:
Interesting, that might work!
But I don't think it's a good idea to keep both conventions, I'll try to use yours.

Pros:
+ I recognize that it has the added benefit that the agent configuration are more visible in the constructor interface, where they probably should be.
+ type hinting can be used

Cons:
- mutables cannot be used as default values for configurations, as they could when defining a config as a dict.
e.g. `class Network(layers=[10, 10, 10])` is forbidden, while default_config = {"layers": [10, 10, 10]} is allowed

Reply:
I see, I think we can do the following for the default mutables:

```python
class Network:
    def __init__(self, layers=None):
           if layers is None:
              layers = [10, 10, 10]
```
or something like this. Then, we can put the default values in the docstring. Some editors, as vscode, show the dosctring and the required parameters by the class as we write the code. So this approach should overall be more user-friendly: 

![image](https://user-images.githubusercontent.com/10488512/100474287-9c6b0900-30e0-11eb-9a3f-b8ae7f40c352.png)

Reply:
Yes, but then the default configuration is more buried and part of the code, rather than at the interface or in a separate dict. Using the doc is perhaps a good idea, but harder to maintain.

But I agree that it is the intended python syntax, so that's a very compelling argument.

My main concern is the ability to save/load agent configurations, with their hierarchical structure, in a text format that can be easily edited.

Reply:
Do we really need this standardised way to save/load agent configurations?

Why can't we just write a config.py file, defining all the parameters/dictionaries required by the agent? 
Then, if we want to keep a backup of the parameters of an experiment, the agent can just copy this file to some output folder.

I like this approach because it makes the code much simpler for the user, and:

* It seems to avoid the issues in the discussion, and makes it easy to have a "transparent" constructor, where we can see all the parameters, instead of passing dictionaries everywhere.

* It is much more flexible and avoids complicated dictionaries as input. For example, I'm writing a code for an agent that takes an UncertaintyEstimator (see rlberry/exploration_tools) as input, and I'm doing the following:
```python
class AgentWithBonus:
    def __init__(self, env, n_episodes, gamma, uncertainty_estimator_fn):
          self.uncertainty_estimator = uncertainty_estimator_fn()
         ...
```
where `uncertainty_estimator_fn` is a function returning one instance of an UncertaintyEstimator.  To create such instance, we need more parameters. 
Writing the constructor as I did above, simplifies a lot the Agent code: we don't need to send an extra dictionary for the agent, so that it creates an instance UncertaintyEstimator using this dictionary.

Then, we can define `uncertainty_estimator_fn` in a `config.py` file easily. 

I'm using the same strategy for algorithms using feature maps, where I send a function `feature_map_fn` to the agent constructor, that can create an instance of a `FeatureMap`.

We could do the same for creating instances of neural nets too.

Notice that this is not incompatible with using `.json` config files, the user can always make a `config.py` that reads a `config.json`, and, after the experiments, save this `config.json` elsewhere as backup.

Reply:
To summarize the above:

I don't think the agent should handle all the parameters of everything it creates. In the example I used, `UncertaintyEstimator` can be a lot of things that implement the same interface: a class that counts the number of visits to each (s, a);  a random network distillation, a density estimator etc.  By doing what I said, the agent doesn't need to worry about this.

Reply:
>Why can't we just write a config.py file, defining all the parameters/dictionaries required by the agent?
> Then, if we want to keep a backup of the parameters of an experiment, the agent can just copy this file to some output folder.

It just seems like a very bad practice to me, I think there's a reason why yml files are used everywhere for configurations. The python code used to store config values in your proposition makes it prone to side effects (no boundary over what this code does), cannot be parsed by another module, compared to another configuration, etc.

> It seems to avoid the issues in the discussion, and makes it easy to have a "transparent" constructor, where we can see all the parameters, instead of passing dictionaries everywhere.

Yes, I agree with this.

> Writing the constructor as I did above, simplifies a lot the Agent code: we don't need to send an extra dictionary to the agent, so that it creates an instance UncertaintyEstimator using this dictionary. Then, we can define uncertainty_estimator_fn in a config.py file easily.

I don't see what using a function brings, you could've sent the constructed UncertaintyEstimator as a parameter directly. But then, you are only breaking the construction of the agent into many separate parts, which seems to make it more complicated for the user, compared to just one call to the Agent contructor.

> Regarding .json config files, the user can always make a config.py that reads a config.json, and, after the experiments, save this config.json elsewhere as backup

Having a reliable and flexible configuration system seems an important feature to me, especially given how hard it is to debug/reproduce RL results. It should not be the responsibility of the user to implement it.

> I don't think the agent should handle all the parameters of everything it creates. In the example I used, UncertaintyEstimator can be a lot of things that implement the same interface: a class that counts the number of visits to each (s, a); a random network distillation, a density estimator etc. By doing what I said, the agent doesn't need to worry about this.

I absolutely agree, and it is also the case if the agent merely forwards to the UncertaintyEstimator (or any subpart) its configuration (e.g.), and it is the UncertaintyEstimator's responsibility to decode it (and fill in missing values using defaults).

Reply:
> It just seems like a very bad practice to me, I think there's a reason why yml files are used everywhere for configurations. The python code used to store config values in your proposition makes it prone to side effects (no boundary over what this code does), cannot be parsed by another module, compared to another configuration, etc.
[...]
> Having a reliable and flexible configuration system seems an important feature to me, especially given how hard it is to debug/reproduce RL results. It should not be the responsibility of the user to implement it.

I agree! A friend recommended the [Sacred](https://github.com/IDSIA/sacred) library, and it can be easily be included in the pipeline as in [this notebook that I added to the examples](https://colab.research.google.com/github/rlberry-py/rlberry/blob/main/notebooks/experimental_pipeline_with_rlberry.ipynb).
`Sacred` can read/write  .json or .yml files to reproduce the experiments and it's very easy to use with `rlberry`.
What to you think ?

> I don't see what using a function brings, you could've sent the constructed UncertaintyEstimator as a parameter directly. But then, you are only breaking the construction of the agent into many separate parts, which seems to make it more complicated for the user, compared to just one call to the Agent contructor.

If we are creating several agents, passing the a constructed UncertaintyEstimator directly would require us to construct several instances (by deepcopy, for example) to send to each agent. So passing a function solves this issue. 

I think it's simpler to pass a function to the constructor than passing nested dictionaries. Also, by doing that, the agent does not need to know which subclass of UncertaintyEstimator it has to instantiate.